### PR TITLE
Future milestone

### PR DESCRIPTION
**Pull request by VAIBHAVJINDAL3012**
# Description

Please provide a detailed description of what was done in this PR
This PR includes future milestone changes which includes the following logic-
1. When we are not able to verify the milestone for any reason, it goes into the future milestone list(assuming Heimdall as source of truth)
2. When we receive any future chain from the peers, we check it against the future list
3. When while whitelisting the milestone, we remove it from the future milestone list if it exist.
# Changes

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)
- [ ] Changes only for a subset of nodes

# Breaking changes

Please complete this section if any breaking changes have been made, otherwise delete it

# Nodes audience

In case this PR includes changes that must be applied only to a subset of nodes, please specify how you handled it (e.g. by adding a flag with a default value...)

# Checklist

- [ ] I have added at least 2 reviewer or the whole pos-v1 team
- [ ] I have added sufficient documentation in code
- [ ] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [ ] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [ ] I have tested this code manually on mumbai
- [ ] I have created new e2e tests into express-cli

### Manual tests

Please complete this section with the steps you performed if you ran manual tests for this functionality, otherwise delete it

# Additional comments

Please post additional comments in this section if you have them, otherwise delete it

### CODE REVIEW

File: common/milestone.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2014 The go-ethereum Authors
```
Comment (manav2401) [inline common/milestone.go:1]: Can we remove this as this is our file? 😅

Comment (VAIBHAVJINDAL3012) [reply common/milestone.go:1]: Yes,Actually It was being used in another implementation. I forgot to remove it.

File: core/rawdb/milestone.go [hunk 2]
```diff
@@ -163,3 +169,53 @@ func ReadLockField(db ethdb.KeyValueReader) (bool, uint64, common.Hash, map[stri
...
+	key := futureMilestoneKey
+
+	enc, err := json.Marshal(futureMilestoneField)
+	if err != nil {
+		log.Error("Failed to marshal the future milestone field struct", "err", err)
+
+		return fmt.Errorf("%w: %v for future milestone field struct", ErrIncorrectFutureMilestoneFieldToStore, err)
+	}
+
+	if err := db.Put(key, enc); err != nil {
```
Comment (JekaMas) [inline core/rawdb/milestone.go:189]: `err =`

File: eth/downloader/whitelist/milestone.go [hunk 3]
```diff
@@ -66,6 +74,14 @@ func (m *milestone) Process(block uint64, hash common.Hash) {
 
 	m.finality.Process(block, hash)
 
+	for i := 0; i < len(m.FutureMilestoneOrder); i++ {
```
Comment (JekaMas) [inline eth/downloader/whitelist/milestone.go:77]: is it just `for _, milestone := range m.FutureMilestoneOrder {`?

File: eth/downloader/whitelist/milestone.go [hunk 4]
```diff
@@ -181,3 +197,58 @@ func (m *milestone) GetMilestoneIDsList() []string {
 func (m *milestone) purgeMilestoneIDsList() {
 	m.LockedMilestoneIDs = make(map[string]struct{})
 }
+
+func (m *milestone) IsFutureMilestoneCompatible(chain []*types.Header) bool {
+
+	chainTipNumber := chain[len(chain)-1].Number.Uint64()
+
+	for i := len(m.FutureMilestoneOrder) - 1; i >= 0; i-- {
```
Comment (JekaMas) [inline eth/downloader/whitelist/milestone.go:205]: a few comments might be useful here

File: eth/downloader/whitelist/milestone.go [hunk 5]
```diff
@@ -181,3 +197,58 @@ func (m *milestone) GetMilestoneIDsList() []string {
...
+
+	if len(m.FutureMilestoneOrder) < m.MaxCapacity {
+		m.enqueueFutureMilestone(num, hash)
+	}
+}
+
+// EnqueueFutureMilestone add the future milestone to the list
+func (m *milestone) enqueueFutureMilestone(key uint64, hash common.Hash) {
+
+	if _, ok := m.FutureMilestoneList[key]; !ok {
```
Comment (JekaMas) [inline eth/downloader/whitelist/milestone.go:232]:
```
if _, ok := m.FutureMilestoneList[key]; ok {
    return
}

log.Debug ...
```

**Review by JekaMas — APPROVED** [review]

File: eth/downloader/whitelist/milestone.go (hunk 3 above)
Comment (VAIBHAVJINDAL3012) [reply eth/downloader/whitelist/milestone.go:77]: Yes

File: eth/downloader/whitelist/milestone.go (hunk 4 above)
Comment (VAIBHAVJINDAL3012) [reply eth/downloader/whitelist/milestone.go:205]: Okay I will add comments here.

File: eth/downloader/whitelist/milestone.go (hunk 5 above)
Comment (VAIBHAVJINDAL3012) [reply eth/downloader/whitelist/milestone.go:232]: I will add the log

**Pull request merged**